Modules/rules.py

```diff
@@ -0,0 +1,107 @@
+
+    Arguments:
+        regex (str):
+            Regular expression to match the command.
+        case_sensitive (bool):
+            Whether to match case-sensitively (using the re.IGNORECASE flag).
+        full_message (bool):
+            If this is True, will try to match against the full message. Otherwise,
+            only the word will be matched against.
+        pass_match (bool):
+            If this is True, the match object will be passed as an argument toward the command
+            function.
+        prefixless (bool):
+            If this is True, the rule can match whether or not the prefix is present. Otherwise, it
+            acts similarly to a command, only being considered if the message starts with our
+            prefix. The prefix itself should not be part of the regex for non-prefixless rules.
+
+    Please note that *regex* can match anywhere within the string, it need not match the entire
+    string. If you wish to change this behaviour, use '^' and '$' in your regex.
+    """
+    def decorator(coro: Callable):
+        if case_sensitive:
+            pattern = re.compile(regex)
+        else:
+            pattern = re.compile(regex, re.IGNORECASE)
+
+        tuple_ = _RuleTuple(pattern, coro, full_message, pass_match)
+        if prefixless:
+            if tuple_ in _prefixless_rules:
+                raise ValueError("rule with the same matching criteria is already registered")
+            _prefixless_rules.append(tuple_)
+        else:
+            if tuple_ in _rules:
+                raise ValueError("rule with the same matching criteria is already registered")
+            _rules.append(tuple_)
+
+        log.info(f"New rule matching '{regex}' case-{'' if case_sensitive else 'in'}sensitively was"
+                 f" created.")
+        return coro
+    return decorator
+
+
+def get_rule(words: List[str], words_eol: List[str],
+             prefixless: bool) -> Tuple[Optional[Callable], tuple]:
+    """
+    Attempt to find a rule in the given dict of patterns and rules.
+
+    Args:
+        words: Words of the message.
+        words_eol: Words of the message, but each including everything to the end of it.
+        prefixless: Whether or not we're looking for prefixless rules.
+
+    Returns:
+        (Callable or None, tuple):
+            2-tuple of the command function and the extra args that it should
+            be called with.
+    """
+    for (pattern, fun, full_message, pass_match) in (_prefixless_rules if prefixless else _rules):
+        if full_message:
+            match = pattern.match(words_eol[0])
+        else:
+            match = pattern.match(words[0])
+
+        if match is not None:
+            if pass_match:
+                return fun, (match,)
+            else:
+                return fun, ()
+    else:
+        return None, ()
+
+
+def clear_rules():
+    global _prefixless_rules, _rules
+    _prefixless_rules = []
+    _rules = []
```

Comment:
`list.clear()` can be called here to clear the rules, this would remove the need for the `global` keyword

Reply:
True, neat idea